fix: update alter table grammar

amanPras opened the pull request:
## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Test A
- [ ] Test B

## Checklist:
- [ ] Each of my commits contains one meaningful change
- [ ] I have performed rebase of my branch on top of the development
- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] I have checked my code and corrected any misspellings

amanPras committed:
fix: update alter table grammar

Signed-off-by: amanprashant <aman.prashant@broadcom.com>

slavek-kucera commented on server/engine/src/main/antlr4/org/eclipse/lsp/cobol/implicitDialects/sql/Db2SqlExecParser.g4:
```diff
@@ -192,10 +192,13 @@ dbs_alter_sequence_loop: (RESTART (WITH INTEGERLITERAL)? | (INCREMENT BY|MINVALU
+                | ENABLE ARCHIVE USE dbs_table_name | DISABLE ARCHIVE | NO KEY LABEL | KEY LABEL dbs_sql_identifier)+;
+dbs_alter_table_add: ADD (COLUMN? dbs_alter_table_coldef | dbs_alter_table_check_constraint | PARTITION (BY dbs_alter_table_partitioning | dbs_alter_table_partition?) |
+                     SYSTEM? VERSIONING USE HISTORY TABLE dbs_alias_name (ON DELETE ADD EXTRA ROW)? | (MATERIALIZED? QUERY)? dbs_alter_table_mq | CLONE dbs_clone_table_name | RESTRICT ON DROP)
+                     | ADD? (dbs_alter_table_unique_constraint | dbs_alter_table_referential)
```
Is the `ADD` really optional?

amanPras replied:
Not as per the [doc](https://www.ibm.com/docs/en/db2-for-zos/13?topic=statements-alter-table), but I am able to compile these sql statements without ADD prefix
For e.g.
```
                     alter TABLE DSN8C10.DEPT
                     CONSTRAINT xyz 
                     UNIQUE (col1, BUSINESS_TIME WITHOUT OVERLAPS)
```

slavek-kucera replied:
It is actually in the doc...
```
The ADD keyword is optional for referential-constraint or unique-constraint if it is the first clause specified in the statement. Otherwise, ADD is required.
```

slavek-kucera commented on server/engine/src/main/antlr4/org/eclipse/lsp/cobol/implicitDialects/sql/Db2SqlExecParser.g4:
```diff
@@ -192,10 +192,13 @@ dbs_alter_sequence_loop: (RESTART (WITH INTEGERLITERAL)? | (INCREMENT BY|MINVALU
+                     SYSTEM? VERSIONING USE HISTORY TABLE dbs_alias_name (ON DELETE ADD EXTRA ROW)? | (MATERIALIZED? QUERY)? dbs_alter_table_mq | CLONE dbs_clone_table_name | RESTRICT ON DROP)
+                     | ADD? (dbs_alter_table_unique_constraint | dbs_alter_table_referential)
+                     ;
+dbs_alter_table_coldef: dbs_column_name_without_alias (dbs_alter_table_bit | dbs_distinct_type_name) (dbs_alter_table_defclause | NOT NULL | dbs_alter_table_check_constraint | common_reference_clause | dbs_alter_table_generated |
```
is the column-definition in alter statement different from the create table case?

amanPras replied:
yes, they are slight different [alter_table_doc](https://www.ibm.com/docs/en/db2-for-zos/13?topic=statements-alter-table) vs [create_table_doc](https://www.ibm.com/docs/en/db2-for-zos/13?topic=statements-create-table)

But, I didn't try to compile all the diff paths.

amanPras replied:
Thanks! updated . Now reusing same column definition for alter table column and create table

slavek-kucera commented on server/engine/src/main/antlr4/org/eclipse/lsp/cobol/implicitDialects/sql/Db2SqlExecParser.g4:
```diff
@@ -215,22 +218,33 @@ dbs_alter_table_bit_timestamp: TIMESTAMP (LPARENCHAR INTEGERLITERAL RPARENCHAR)?
                     (ENABLE|DISABLE) QUERY OPTIMIZATION (MAINTAINED BY (SYSTEM|USER))?);
-dbs_alter_table_alter: ALTER (COLUMN? dbs_alter_table_colalt | PARTITION INTEGERLITERAL dbs_alter_table_partition | MATERIALIZED? QUERY dbs_alter_table_mq);
-dbs_alter_table_colalt: dbs_column_name_without_alias (DROP DEFAULT | SET (DATA TYPE dbs_alter_table_bit (INLINE LENGTH INTEGERLITERAL)? | dbs_alter_table_defclause | INLINE LENGTH INTEGERLITERAL | dbs_alter_table_idalt |
+dbs_alter_table_alter: ALTER (COLUMN? dbs_alter_table_colalt | dbs_alter_table_partition | MATERIALIZED? QUERY dbs_alter_table_mq | PARTITIONING TO PARTITION BY dbs_alter_table_partitioning);
```
The alter partition is still a valid command.

slavek-kucera commented on server/engine/src/main/antlr4/org/eclipse/lsp/cobol/implicitDialects/sql/Db2SqlExecLexer.g4:
```diff
@@ -535,6 +535,7 @@ PARAMETER : P A R A M E T E R;
 PART : P A R T;
 PARTITION : P A R T I T I O N;
 PARTITIONED: P A R T I T I O N E D;
+PARTITIONING: P A R T I T I O N I N G;
```
Does this new token need to be added to `dbs_special_name`?

amanPras replied:
added at line 1598 of parser

amanPras committed:
chore: use compilable code for use case test

Signed-off-by: amanprashant <aman.prashant@broadcom.com>

amanPras committed:
fix: reuse same column definition for alter table column and create table

Signed-off-by: amanprashant <aman.prashant@broadcom.com>

slavek-kucera commented on server/engine/src/main/antlr4/org/eclipse/lsp/cobol/implicitDialects/sql/Db2SqlExecParser.g4:
```diff
@@ -215,22 +216,29 @@ dbs_alter_table_bit_timestamp: TIMESTAMP (LPARENCHAR INTEGERLITERAL RPARENCHAR)?
 dbs_alter_table_defclause: WITH? DEFAULT (dbs_constant | SESSION_USER | USER | CURRENT SQLID | NULL | dbs_cast_function_name LPARENCHAR (dbs_constant | SESSION_USER | USER | CURRENT SQLID | NULL) RPARENCHAR)?;
-dbs_alter_table_check: (CONSTRAINT dbs_constraint_name)? CHECK LPARENCHAR dbs_search_condition RPARENCHAR;
-common_reference_clause: REFERENCES dbs_table_name (LPARENCHAR dbs_column_name_without_alias (PERIOD BUSINESS_TIME)? (dbs_comma_separator dbs_column_name (PERIOD BUSINESS_TIME)?)* RPARENCHAR)? (ON DELETE (RESTRICT | NO ACTION | CASCADE |
+common_reference_clause: REFERENCES dbs_table_name (LPARENCHAR dbs_column_name_without_alias  (dbs_comma_separator dbs_column_name)* RPARENCHAR)? (ON DELETE (RESTRICT | NO ACTION | CASCADE |
```
dbs_column_name -> dbs_column_name_without_alias?
```suggestion
common_reference_clause: REFERENCES dbs_table_name (LPARENCHAR dbs_column_name_without_alias  (dbs_comma_separator dbs_column_name_without_alias)* RPARENCHAR)? (ON DELETE (RESTRICT | NO ACTION | CASCADE |
```

amanPras committed:
Update server/engine/src/main/antlr4/org/eclipse/lsp/cobol/implicitDialects/sql/Db2SqlExecParser.g4

Co-authored-by: slavek-kucera <53339291+slavek-kucera@users.noreply.github.com>

amanPras committed:
chore: revert changes for common_reference_clause rule

revert changes for common_reference_clause rule till we analyze that PERIOD BUSINESS_TIME is not supported

Signed-off-by: amanprashant <aman.prashant@broadcom.com>
